### PR TITLE
Fix BYOL-A implementation

### PR DESCRIPTION
**Pull request by daisukelab**
Hi, let me make a request to fix the BYOL-A implementation. We made the followings:

1. Fixed the usage regarding the calculation of normalization statistics.

It is now required to calculate the normalization statistics beforehand.

    python run_downstream.py -m train -n byola_calcnorm_vc1 -u byol_a_calcnorm -d voxceleb1 -k ./AudioNTT2020-BYOLA-64x96d2048.pth

This will output:

    ** Running Norm has finished updates over 10000 times, using the following stats from now on. ***
    mean,std=-8.907230377197266,4.892485618591309
    *** Please use these statistics in your model. EXIT... ***

Then we can train and test.

    python run_downstream.py -m train -n my_byol_a_vc1_1 -u byol_a -d voxceleb1 -o "config.optimizer.lr=1e-3" -k ./AudioNTT2020-BYOLA-64x96d2048.pth,-8.907230377197266,4.892485618591309
    python run_downstream.py -m evaluate -e result/downstream/my_byol_a_vc1_1/dev-best.ckpt

2. Enhanced to collect features from all the layers.

Now the model collects features from all 5 layers (although it doesn't push the performance much).

And some small fixes. Thanks!

### CODE REVIEW

**Comment by daisukelab** [comment]
Let me reopen this. Please let me know if anything needs to be corrected.

**Comment by leo19941227** [comment]
Oh! I did not notice this PR. Sorry for that, that me review on this later.

File: s3prl/upstream/byol_a/expert.py [hunk 1]
```diff
@@ -11,9 +11,9 @@
```
Comment (leo19941227) [inline s3prl/upstream/byol_a/expert.py:11]: https://github.com/s3prl/s3prl/blob/4f5ad57dbff1b10234b0cb1cc35f03b64719a60a/s3prl/upstream/byol_a/expert.py#L2-L7

Feel free to add your name and the institution as the author.

File: s3prl/upstream/byol_a/expert.py [hunk 2]
```diff
@@ -29,43 +29,36 @@ def __init__(
...
-        self.normalizer = PrecomputedNorm(stats)
+        self.to_logmelspec = LogMelSpectrogram()
+        if norm_mean is None or norm_std is None:
+            print('  ** CAUTION **')
+            print('  This is a run for calculating statistics of the downstream task using RunningNorm and will exit in the middle of training. **')
+            print('  ** CAUTION **')
+            self.normalizer = RunningNorm(epoch_samples=10_000, max_update_epochs=1, axis=[0, 1, 2]) # Use single scalar mean/std values.
+        else:
+            print(f'*** Using normalization statistics: mean={norm_mean}, std={norm_std} ***')
+            self.normalizer = lambda x: (x - norm_mean) / norm_std
```
Comment (leo19941227) [inline s3prl/upstream/byol_a/expert.py:47]: Let's use the logger package instead of printing.

```python3
import logging
logger = logging.getLogger(__name__)

logger.info("some information")
```

File: s3prl/upstream/byol_a/README.md [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+- 1024-d feature: https://github.com/nttcslab/byol-a/raw/master/pretrained_weights/AudioNTT2020-BYOLA-64x96d1024.pth
+- 512-d feature: https://github.com/nttcslab/byol-a/raw/master/pretrained_weights/AudioNTT2020-BYOLA-64x96d512.pth
+
+### 1. Configuring the number of feature dimensions
+
+Edit your `config.yaml` to match your weight.
+
+    # Dimensions of feature representations.
+    feature_d: 2048
+
```
Comment (leo19941227) [inline s3prl/upstream/byol_a/README.md:27]: Is it possible to hide this step from simply inferring this information from checkpoint?
We may consider to store the config into the checkpoint?

File: s3prl/upstream/byol_a/byol_a.py [hunk 4]
```diff
@@ -138,3 +161,218 @@ def forward(self, x):
...
+        if len(self.ema_mean) < self.max_update:
+            self.ema_mean.put(image)
+            self.ema_var.update_mean(self.ema_mean())
+            self.ema_var.put(image)
+            self.mean = self.ema_mean()
+            self.std = torch.clamp(self.ema_var.std(), torch.finfo().eps, torch.finfo().max)
+        elif not self.reported:
+            self.reported = True
+            print(f'*** Running Norm has finished updates over {self.max_update} times, using the following stats from now on. ***\n  mean,std={float(self.mean.view(-1))},{float(self.std.view(-1))}\n')
+            print(f'*** Please use these statistics in your BYOL-A. EXIT... ***\n')
```
Comment (leo19941227) [inline s3prl/upstream/byol_a/byol_a.py:371]: Let's use the logging package with
```python3
logger.info
```

File: s3prl/upstream/byol_a/expert.py [hunk 5]
```diff
@@ -29,43 +29,36 @@ def __init__(
...
-            -5.4919195,
-            5.0389895,
-        ]  # FIXME: should use downstream dataset statistics
-        self.normalizer = PrecomputedNorm(stats)
+        self.to_logmelspec = LogMelSpectrogram()
+        if norm_mean is None or norm_std is None:
+            print('  ** CAUTION **')
+            print('  This is a run for calculating statistics of the downstream task using RunningNorm and will exit in the middle of training. **')
+            print('  ** CAUTION **')
+            self.normalizer = RunningNorm(epoch_samples=10_000, max_update_epochs=1, axis=[0, 1, 2]) # Use single scalar mean/std values.
```
Comment (leo19941227) [inline s3prl/upstream/byol_a/expert.py:44]: Is `10_000` a typo?

File: s3prl/upstream/byol_a/expert.py [hunk 6]
```diff
@@ -80,50 +73,11 @@ def forward(self, wavs):
...
-        trimmed_hs = []
-        for h in [repre]:
-            max_h_len = len(range(0, max_wav_len, self.get_downsample_rates()))
-            h = h[:, :max_h_len, :]
-            trimmed_hs.append(h)
-
+        self.to_logmelspec.to(wavs[0].device)
+        wavs = pad_sequence(wavs, batch_first=True)
+        features = self.normalizer(self.to_logmelspec(wavs)).unsqueeze(1) # (B, F, T) -> (B, 1, F, T)
+        layered_features = self.model.by_layers(self.model(features, layered=True)) # [(B, T, D)] x 5
```
Comment (leo19941227) [inline s3prl/upstream/byol_a/expert.py:79]: I have a quick question just to make sure, is the `(B, T, D)` here represents `batch_size`, `number of frames` and `hidden_size`?
If so, is the `number of frames` are in order, so that the later frames always represent the later samples of the input waveform?

**Review by leo19941227 — COMMENTED** [review]
Hi @daisukelab !

Thanks for the PR!
Overall I am good with the current design, while having some minor concerns.
Please take a look and let me know what you think, thanks!

A general feedback is that we are switching from printing to the `logging` package, so please help make sure to use `logger.info` to replace `print` in the current PR.

**Comment by leo19941227** [comment]
Also, I just find that you said you plan to add your new model Masked Modeling Duo (sorry again for not noticing that activity). Let me know if you are still planning to add that so we can schedule work on it after this PR.

File: s3prl/upstream/byol_a/expert.py (hunk 1 above)
Comment (daisukelab) [reply s3prl/upstream/byol_a/expert.py:11]: Thanks, but I think I should refrain from touching on this.

File: s3prl/upstream/byol_a/expert.py (hunk 2 above)
Comment (daisukelab) [reply s3prl/upstream/byol_a/expert.py:47]: OK, I will follow it.

File: s3prl/upstream/byol_a/README.md (hunk 3 above)
Comment (daisukelab) [reply s3prl/upstream/byol_a/README.md:27]: Yes, I will try.

File: s3prl/upstream/byol_a/expert.py (hunk 6 above)
Comment (daisukelab) [reply s3prl/upstream/byol_a/expert.py:79]: Yes, your guesses are right. BYOL-A is a simple CNN that convolves spectrogram frames in order and keeps the order in the output features.

**Comment by daisukelab** [comment]
Thanks for your question about Masked Modeling Duo. We plan to release the code for SUPERB in May or June if anything goes well with Interspeech.

File: s3prl/upstream/byol_a/expert.py (hunk 5 above)
Comment (leo19941227) [reply s3prl/upstream/byol_a/expert.py:44]: Oh! Thanks for the clarification, I did not know that.
Indeed I could successfully run the model without changing this line but I did not know why lol

File: s3prl/upstream/byol_a/expert.py (hunk 6 above)
Comment (leo19941227) [reply s3prl/upstream/byol_a/expert.py:79]: Cool!

**Comment by leo19941227** [comment]
> Thanks for your question about Masked Modeling Duo. We plan to release the code for SUPERB in May or June if anything goes well with Interspeech. (We have one more thing with the M2D :) )

Sounds cool! Hope everything goes well.

**Comment by daisukelab** [comment]
@leo19941227  Hi, I made all the changes needed. Could you take a look again, please? Thanks!

**Comment by leo19941227** [comment]
Sure! Sorry I am out for a short vacation this weekend. I can take a look on Monday. Thanks for the quick update!

**Comment by daisukelab** [comment]
No problem, I'm on vacation, too ;) (submission complete!)
